docs/proposals/2020-08-12-Collection_Query_Backfill.md

```diff
@@ -0,0 +1,72 @@
+# Collection Query Backfill
+
+## Problem
+
+Backfilling Collections with query results is a complex process that requires implementation across multiple systems.
+This document proposes a first solution to achieve the near-term deliverable.
+
+## Proposal
+
+In order to limit scope on both the rendering and publishing sides, we only allow specifying a single section or tag. The Collections API would be updated to support this via the API and presumably UI as well.
+
+### Example
+
+This document propose to implement the first iteration of query backfill.
+In this first step we would update the ANS spec to allow specifying a single section or tag.
+A Subsequent releases will be made to support multiple tags and sections.
+The current proposed schema is setting the contract for the first release and all subsequent releases, as such we call so lets call those fields tags and sections and use arrays for future proofing.
+
+```JSON
+{
+  "type": "collection",
+  "dynamic_items": {
+    "type": "sections",
+    "ids": ["/politics"]
+  },
+}
+```
+
+```JSON
+{
+  "type": "collection",
+  "dynamic_items": {
+    "type": "tags",
+    "ids": ["basketball"]
+  },
+  ...
+}
+```
+
+When Content API fetches collections with the dynamic_items property (or whatever we want to name it), it will merge them in from a query.
+
+## Concerns
+
+### How are website restrictions handled?
+
+Collections are currently limited to a single website, but there are product plans to expand that to multiple websites. Sections exist on an individual website. From a technical perspective, tags are a much cleaner solution, but we could implement a rule where a section backfill is ignored on websites other than the primary website. I think that’s a bad experience, though.
+
+### How are dynamically filled items marked inside the collection?
+
+We could add is_dynamic_item under additional_properties, but I’m open to alternatives.
+
+### How does Content API backfill the collections?
+
+#### Section-based example query:
+
+`https://api.washpost.arcpublishing.com/content/v4/search/published?website=washpost&q=type:story+AND+taxonomy.sites._id:%22/politics%22&sort=display_date:desc&size=10`
+
+#### Tag-based example query:
+
+`https://api.washpost.arcpublishing.com/content/v4/search/published?website=washpost&q=type:story+AND+taxonomy.tags.slug:obama&sort=display_date:desc&size=10`
+
+### Are there limits on the number of dynamically-filled items?
+
+I suggest limiting it to the same as Content API’s general limit, which would be a total of 100 items.
```

Comment:
I imagine that customers might expect any limit to respect the collection's `cap`, which currently defaults to its maximum value of `500`. If an organization were to have a collection with a cap of 50 and 20 hand-curated content items, they would probably expect a maximum of 30 additional stories to be returned by the query backfill.

Reply:
Hm, but how does the Collection cap interact with the result set size in the Content API request?

Reply:
Oh, really? I read it how you originally read it

Reply:
`https://api.washpost.arcpublishing.com/content/v4/search/published?website=washpost&q=type:story+AND+taxonomy.sites._id:%22/politics%22&sort=display_date:desc&size=10`

Given a Collection with five manually curated items and a cap of 20, my expectation is that the Content API would return 10 stories, including five backfilled from the Politics section.

Reply:
I thought

> Oh, really? I read it how you originally read it

I thought so too until I saw the implementation, so this probably merits clarification.

Reply:
yes, the first implementation will limit the number of tags sections to one, which will be restricted to the application layer. A second iteration will expand the same number to 100